Support for quadrant degree bearing readings

echarlie opened the pull request:
At this point, I think I've correctly handled tests, english error messages, and the read_quadrants function s.t. it won't break under multi-value circumstances (if that ever gets implemented). I've done rudimentary testing of the error messages as well---I'm not sure I entirely understand the control flow, but I think it's been implemented correctly.

echarlie committed:
first cut at quadrant support

echarlie committed:
try to add units support. something's messed up

echarlie committed:
table values must be alphabetical...

echarlie committed:
use a const multiplier instead of literal 90/180/270/360, in case someone wants grads

echarlie committed:
fix some bits to be more clear, and eliminate some faffing about

echarlie committed:
split quadrant handling into seperate function, with some boilerplate from read_number

echarlie committed:
start working on basic tests

echarlie committed:
fix errors with cardinal directions

echarlie committed:
wrap quadrant handling in read_numeric_multi

echarlie committed:
better handle omit

echarlie committed:
ugh, messages

ojwb committed:
Merge branch 'master' into HEAD

ojwb committed:
Clean up indentation and other whitespace

ojwb committed:
Update units_to_msgno table

Fixes cavern.tst testcase "back".

ojwb committed:
Renumber new message

Use an unused message number within the existing range rather
then using a new number well above any currently used.  This avoids
extra zero padding in all the compiled message files, which in turn
increases start up time.

ojwb commented:
I've merged master to the branch as a bug fix I made on master was conflicting (only trivially but in a way git couldn't resolve automatically).

I've also fixed a testcase which was failing due to the changes and done a bit of clean up.

On my laptop, surfequate fails due to getting `-0.00` instead of `0.00` in DXF export, but I'm also seeing that on master.  Might be an i386-specific issue, due to excess precision from 387 FP instructions so I've pushed the changes to see what CI shows.

ojwb commented:
CI passed on Linux at least, so this indeed seems to be architecture-specific and unrelated to this patch.  It needs addressing, but isn't a blocker for merging this.

ojwb committed:
Adjust new testcase

Make the expected result a pos file like existing cavern.tst testcases
and annotate so unexpected warnings will cause the testcase to fail.

ojwb committed:
More testcase adjustments

Test "S" works, check lower and mixed case versions work, and break the
loop so we more clearly test the bearings are correctly interpreted.

ojwb committed:
Ship new testcase files

ojwb committed:
Improve reporting of bad quadrant bearings

ojwb commented:
I've improved the test coverage, and especially there's now a test of all the error cases I could think of.  That uncovered some cases that were getting a non-helpful error and I've ended up adjusting the parsing so that all the cases at least get the "Expecting quadrant bearing" error message.  This is perhaps a bit generic, but at least clearly indicates the field that's wrong, which should be enough for the user to be able to sort out the issue.

I'm wondering if something like `S99E` should be a warning rather than an error, akin to a bearing of `369` degrees, but maybe an error is better.

Aside from that, I should probably do a final look over but this is looking good.

ojwb committed:
Fix handling of plumbs with quadrant bearings

ojwb committed:
Improve quadrant bearing errors

Give a "Suspicious compass reading" error if the prefix and suffix
are valid, but the number is > 90
